# Authors:

# Assignment 3

In [ ]:
# import libraries
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

# Question 1: 1.Load the Breast Cancer dataset from sklearn.datasets.

In [27]:
# Load the dataset
breast_cancer_dataset = sklearn.datasets.load_breast_cancer()

print(f'Breast Cancer Dataset loaded with {len(breast_cancer_dataset.data)} samples and {len(breast_cancer_dataset.feature_names)} features.')

Breast Cancer Dataset loaded with 569 samples and 30 features.


# Question 1 explanation

- for question 1, i loaded the dataset using sklearn library's datasets function

# Question 2: Assign the features to a variable called X and the class labels to variable called y. Hint: look at .keys(), .target_names, .feature_names, etc.

In [30]:
# Assign features to X and class labels to y
print(f"The keys of the breast_cancer_dataset are: {breast_cancer_dataset.keys()}")
X = breast_cancer_dataset.data
y = breast_cancer_dataset.target
print(f'\nFeatures (X) are {breast_cancer_dataset.feature_names} with shape: {X.shape}')
print(f'\nClass labels (y) are {breast_cancer_dataset.target_names} with shape: {y.shape}')

The keys of the breast_cancer_dataset are: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

Features (X) are ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension'] with shape: (569, 30)

Class labels (y) are ['malignant' 'benign'] with shape: (569,)


# Question 2 explanation

- for question 2, i first looked into what the breast cancer dataset contains by using the function .keys()
- then, i extracted the X and y datasets using .data and .target functions
- finally i reported the feature and label names as well as the dimension of each dataset

# Question 3: Split the data into train and test sets. Hint: use train_test_split().

In [32]:
# Split the data into training and testing sets. Print the results
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(f'Training features shape: {X_train.shape}')
print(f'Testing features shape: {X_test.shape}')
print(f'Training labels shape: {y_train.shape}')
print(f'Testing labels shape: {y_test.shape}')

Training features shape: (455, 30)
Testing features shape: (114, 30)
Training labels shape: (455,)
Testing labels shape: (114,)


# Question 3 explanation:

- i first needed to utilize function train_test_split with 80-20 split and set the random seed to reproduce the results

- after I split the data into train and test sets, I reported the shapes for each dataset

- as can be seen from the data shapes, the train dataset has the most data for the model to learn from while test dataset is a smaller subset but has enough points to make predictions from

# Question 4: Choose your favourite classifier and see if you can classify the benign and cancer cases. For a list of classifiers, see: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning.

In [ ]:
# Try random forest
# Initalize the classifier
clf = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the classifier to the training data and make predictions on the test data
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

# Calculate accuracy and report
accuracy = accuracy_score(y_test, y_pred)
print(f'Random Forest Classifier Accuracy: {accuracy:.2f}')

# Try support vector machine
# Initalize the classifier
svc = SVC(kernel='linear', random_state=42)

# Fit the classifier to the training data and make predictions on the test data
svc.fit(X_train, y_train)
y_pred_svc = svc.predict(X_test)

# Calculate accuracy and report
accuracy_svc = accuracy_score(y_test, y_pred_svc)
print(f'Support Vector Machine Classifier Accuracy: {accuracy_svc:.2f}')


Random Forest Classifier Accuracy: 0.96
Support Vector Machine Classifier Accuracy: 0.96


# Explanation for question 4

- i wanted to experiment with random forest and support vector machine

- for each of the classifiers, my first step was to define the classifiers using corresponding functions from sklearn

- then, i trained the initialized classifiers using train dataset

- then, i tested the final classifier using test dataset and reported the accuracy


# Explanation for how random forest classifier works

- i used the scikit learn documentation (link: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
) which provides the explanation for random forest classifier

- according to scikit learn, the random forest classifier is an ensemble method and a meta estimator, meaning it combines multiple models to make a final prediction

- random forest fits a specific number of decision trees (100 in our case) using different subsamples of the given dataset, which is the training dataset in this experiment

- random forest helps control overfitting by averaging the predictions from all decision trees, which also improves the overall predictive accuracy

# Explanation for how support vector machine works

- i used the scikit learn documentation (link: https://scikit-learn.org/stable/modules/svm.html), to also explain the support vector machine classifier. 

- support vector machine, or svm, is a supervised learning algorithm used for classification tasks. 

- svm works by finding a decision boundary that best separates different classes in the dataset. this decision boundary is called a hyperplane.

- the goal of svm is to maximize the margin between the classes and the hyperplane. 

- only a few data points close to the boundary affect the model. 

- these points are called support vectors. by focusing on these support vectors, svm can generalize well and make accurate predictions on unseen data.

# Question 5: Compute the confusion matrix, precision, and recall. Hint: use confusion_matrix(), and classification_report().

In [ ]:
print("Evaluation for Random Forest Classifier:")
# Compute confusion matrix and classification report. Print the results
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)
print('Classification Report:')
print(class_report)



Evaluation for Random Forest Classifier:
Confusion Matrix:
[[40  3]
 [ 1 70]]
Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114



# Explanation for question 5

- i chose to compute the confusion matric and classification report for the random forest

- i computed confusion matrix using the corresponding function in sklearn using test dataset and prediction dataset

- i computed classification report using corresponding function in sklearn using test dataset and prediction dataset

- finally, i reported the results

# Discuss the results

- the random forest classifier achieved a high overall accuracy of 0.96, showing strong performance on the test dataset

- the confusion matrix shows that most samples were classified correctly, with only a few misclassifications

- for class 0-malignant, the model has high precision but slightly lower recall, meaning a few instances were misclassified

- for class 1-benign, both precision and recall are very high, so the model predicts this class very reliably

- overall, the results suggest that the random forest classifier is effective and balanced across both classes